File: mindflow/src/requests/index_references.rs
// This module contains the request to the Mindflow server to index references.

use serde::{Serialize};
use reqwest::{Client};
use std::process;

use crate::requests::status::error::ErrorResponse;
use crate::utils::config::{CONFIG};
use crate::utils::reference::Reference;
       
#[derive(Serialize)]
pub(crate) struct IndexReferencesRequest {
    pub(crate) references: String,
    pub(crate) auth: String
}

impl IndexReferencesRequest {
    pub fn new(references: Vec<Reference>) -> IndexReferencesRequest {
        IndexReferencesRequest {
            references: serde_json::to_string(&references).unwrap_or_else(|_| {
                println!("Error: Could not serialize hashes to JSON.");
                process::exit(1);
            }),
            auth: CONFIG.get_auth_token()
        }
    }
}

// Send a request to the Mindflow server to index references.
pub(crate) async fn request_index_references(client: &Client, unindexed_references: Vec<Reference>) {
    let index_reference_request = IndexReferencesRequest::new(unindexed_references);

    let url = format!("{}/index", CONFIG.get_api_location());
    let res = match client.post(&url).json(&index_reference_request).send().await {
        Ok(res) => res,
        Err(_e) => { process::exit(1) },
    };
    
    // match status code
    match res.status().as_u16() {
        200 => {
            log::debug!("References indexed.");
        }
        status   => {
            let error: ErrorResponse = res.json().await.unwrap();
            println!("Error: {} - {}", status, error.msg);
            process::exit(1);
        }
    };
}

File: mindflow/requests/index_references.py
"""
Request to index references.
"""

import json
from typing import List
import requests

from mindflow.utils.config import Config
from mindflow.utils.reference import Reference


def request_index_references(
    resolved_references: dict[str, Reference], unindexed_hashes: List[str]
):
    """
    This function makes a post request to the backend to index the unindexed references.
    """
    if len(unindexed_hashes) == 0:
        return
    unindexed_references: List[Reference] = [
        vars(resolved_references[unindexed_reference])
        for unindexed_reference in unindexed_hashes
    ]
    response = requests.post(
        f"{Config.API_LOCATION}/index",
        json={"references": json.dumps(unindexed_references), "auth": Config.AUTH},
        timeout=10,
    )
    if response.status_code != 200:
        print(f"Error: {response.status_code} {response.text}")
